circuit-construction/src/lib.rs

```diff
@@ -101,24 +105,47 @@ impl Cycle for FqInner {
     type OuterProj = <Other as AffineCurve>::Projective;
 }
 
-#[derive(Hash, Eq, PartialEq, Debug, Clone, Copy)]
-pub struct Var<F> {
+type Ref<Sys> = Rc<RefCell<Sys>>;
+
+/// A variable in the circuit,
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+pub struct Var<Sys, F>
+where
+    Sys: Clone,
+{
+    /// A unique identifier for this variable.
     pub index: usize,
+    /// ends up holding a value when generating the witness.
     pub value: Option<F>,
+    /// A pointer to the system.
+    pub sys: Sys,
```

Comment:
If this is supposed to be a pointer, make it one. The `Arc<Rc<Sys>>` should be an implementation detail.